### PR TITLE
unix: call setgoups before calling setuid/setgid

### PR DESCRIPTION
**Pull request by saghul**
Partial fix for #1093

Shamelessly taken from: https://github.com/mozilla/rust/pull/12085

/cc @indutny

If you have a few milliseconds to spare, it'd be great to get your input, @bnoordhuis.

As for the dropping caps part, I'll need to look into that, doesn't look anywhere straightforward :-/

### CODE REVIEW

File: src/unix/process.c [hunk 1]
```diff
@@ -330,6 +330,17 @@ static void uv__process_child_init(const uv_process_options_t* options,
     _exit(127);
   }
 
+  if (options->flags & (UV_PROCESS_SETUID|UV_PROCESS_SETGID)) {
```
Comment (indutny) [inline src/unix/process.c]: Please add space between: `...SETUID | ...SETGID`

File: src/unix/process.c [hunk 2]
```diff
@@ -330,6 +330,17 @@ static void uv__process_child_init(const uv_process_options_t* options,
...
+    /* When dropping privileges from root, the `setgroups` call will
+     * remove any extraneous groups. If we don't call this, then
+     * even though our uid has dropped, we may still have groups
+     * that enable us to do super-user things. This will fail if we
+     * aren't root, so don't bother checking the return value, this
+     * is just done as an optimistic privilege dropping function.
+     */
+    setgroups(0, NULL);
```
Comment (indutny) [inline src/unix/process.c:343]: Perhaps, clear errno?

Comment (indutny) [inline src/unix/process.c:343]: I mean restore.

**Comment by indutny** [comment]
Minor style nit, otherwise LGTM. But yeah, having @bnoordhuis opinion on this would be totally cool!

File: src/unix/process.c (hunk 2 above)
Comment (saghul) [inline src/unix/process.c:343]: I thought about it, but we don't check for it unless any of the remaining function calls fail, so does it matter much? Or is it just good practice? I could just set errno to 0 if setgroups fails.

Comment (indutny) [inline src/unix/process.c:343]: Better just restore it, yeah it is just for the sake of style.

**Comment by saghul** [comment]
Force pushed addressing feedback, thanks @indutny!

File: src/unix/process.c [hunk 3]
```diff
@@ -330,6 +330,20 @@ static void uv__process_child_init(const uv_process_options_t* options,
...
+     * remove any extraneous groups. If we don't call this, then
+     * even though our uid has dropped, we may still have groups
+     * that enable us to do super-user things. This will fail if we
+     * aren't root, so don't bother checking the return value, this
+     * is just done as an optimistic privilege dropping function.
+     */
+    int saved_errno;
+    saved_errno = errno;
+    setgroups(0, NULL);
+    errno = saved_errno;
```
Comment (bnoordhuis) [inline src/unix/process.c:344]: Can be shortened to `SAVE_ERRNO(setgroups(0, NULL))` but apart from that LGTM.

**Comment by bnoordhuis** [comment]
Change itself LGTM but as food for thought, you may want to consider adding an API that lets the user retain some or all of the ancillary groups.  Ditto for capabilities.

File: src/unix/process.c (hunk 3 above)
Comment (saghul) [inline src/unix/process.c:344]: Thanks Ben, will fix.

**Comment by saghul** [comment]
Thanks guys! Landed as https://github.com/joyent/libuv/commit/66ab38918c911bcff025562cf06237d7fedaba0c. I also created https://github.com/joyent/libuv/issues/1106 to track the Linux capabilites part and the API suggesting by Ben.

**Pull request closed without merging**